framework/src/main/java/org/checkerframework/common/returnsreceiver/ReturnsReceiverAnnotatedTypeFactory.java

```diff
@@ -0,0 +1,93 @@
+import org.checkerframework.common.returnsreceiver.qual.UnknownThis;
+import org.checkerframework.framework.type.AnnotatedTypeFactory;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.typeannotator.ListTypeAnnotator;
+import org.checkerframework.framework.type.typeannotator.TypeAnnotator;
+import org.checkerframework.javacutil.AnnotationBuilder;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/** The type factory for the Returns Receiver Checker. */
+public class ReturnsReceiverAnnotatedTypeFactory extends BaseAnnotatedTypeFactory {
+
+    /**
+     * The {@code @}{@link This} annotation. The field is package visible due to a use in {@link
+     * ReturnsReceiverVisitor}
+     */
+    final AnnotationMirror THIS_ANNOTATION;
+
+    /**
+     * Create a new {@code ReturnsReceiverAnnotatedTypeFactory}.
+     *
+     * @param checker the type-checker associated with this factory
+     */
+    public ReturnsReceiverAnnotatedTypeFactory(BaseTypeChecker checker) {
+        super(checker);
+        THIS_ANNOTATION = AnnotationBuilder.fromClass(elements, This.class);
+        this.postInit();
+    }
+
+    @Override
+    protected Set<Class<? extends Annotation>> createSupportedTypeQualifiers() {
+        return getBundledTypeQualifiers(BottomThis.class, UnknownThis.class, This.class);
+    }
+
+    @Override
+    protected TypeAnnotator createTypeAnnotator() {
+        return new ListTypeAnnotator(
+                new ReturnsReceiverTypeAnnotator(this), super.createTypeAnnotator());
+    }
+
+    /** A TypeAnnotator to add the {@code @}{@link This} annotation. */
+    private class ReturnsReceiverTypeAnnotator extends TypeAnnotator {
+
+        /**
+         * Create a new ReturnsReceiverTypeAnnotator.
+         *
+         * @param typeFactory the {@link AnnotatedTypeFactory} associated with this {@link
+         *     TypeAnnotator}
+         */
+        public ReturnsReceiverTypeAnnotator(AnnotatedTypeFactory typeFactory) {
+            super(typeFactory);
+        }
+
+        @Override
+        public Void visitExecutable(AnnotatedTypeMirror.AnnotatedExecutableType t, Void p) {
+
+            // skip constructors, as we never need to add annotations to them
+            if (t.getElement().getKind() == ElementKind.CONSTRUCTOR) {
+                return super.visitExecutable(t, p);
+            }
+
+            AnnotatedTypeMirror returnType = t.getReturnType();
+
+            // If any FluentAPIGenerator indicates the method returns this,
+            // add an @This annotation on the return type.
+            if (FluentAPIGenerator.check(t)) {
+                if (!returnType.isAnnotatedInHierarchy(THIS_ANNOTATION)) {
+                    returnType.addAnnotation(THIS_ANNOTATION);
+                }
+            }
+
+            // If return type is annotated with @This, add @This annotation
+            // to the receiver type.
```

Comment:
Maybe add the note about the automatic defaulting here also. Maybe file a follow-up issue to keep track of method reference handling and the false positives.

Reply:
Do we need an issue separate from #2931?

Reply:
Now that there is a reference to that issue, this is fine.

Reply:
Oh, there is none here. Can you add one, to make it easier to find all places that should be touched once #2931 is fixed?

Reply:
It's there:  https://github.com/typetools/checker-framework/pull/3199/files#diff-4fedb1e93f819c78e6268bc5f23478b1R82-R83  The GitHub UI is confusing with outdated views

Reply:
Oh, I specifically tried to look on a reload and didn't see it... thanks, so that's good then.